### DISCUSSION TITLE
Vector lite

### DISCUSSION THREAD

**Discussion by bmallya** [vectordotdev/vector#9010] [General]
Hello, 
I was wondering if anyone has attempted to build a lite version for vector given that only a few of sources/sinks/transforms will be used. 
For example,  Given a case where the _sources_ needed could be http, socket, files 
 while _sinks_ are  vector , file along with the _transforms_, how do we 

1. Identify the lib dependencies that are not required
2. Test the finally compiled binary to ensure that it is good.

I have tried to remove the sources/sinks/transforms that are not required along with any dependencies that are mentioned in Cargo.toml file. This I then verified by compiling vector and then generating dummy data and forwarding from one instance to another instance.

Ex: I removed the following source and also rusto, rusto_s3 from Cargo.toml 
sources-aws_s3 = ["rusoto", "rusoto_s3", "rusoto_sqs", "semver", "uuid"]
But I do not know if there are any other indirect dependencies. 

Attached is a cargo.toml file that I had used (with some sources/sinks/transforms and dependencies 
[vectorCargo.txt](https://github.com/timberio/vector/files/7100987/vectorCargo.txt)
commented). With this file, I was able to get vector of size 77 MB from the default ~145 MB for v0.15.1

**Comment by jszwedko** [comment]
Hi @bmallya !

We actually do enable custom, slimmer, builds through having a Cargo feature flags for each component. For example, if you only needed the components you mentioned, you could do:

```
cargo build --release --no-default-features --features 'sources-file,sources-http,sources-socket,transforms,sinks-vector'
```

If you only need a few select transforms, you can also replace `transforms` (which will include all) with things like `transforms-remap` to only include the `remap` transform.

Hopefully this helps!

**Reply by bmallya** [reply]
Thank you  @jszwedko for the quick response. With `cargo build --release  --no-default-features --features 'sources-exec,sources-file,sources-generator,sources-http,sources-internal_logs,sources-socket,sources-stdin,sources-syslog,sources-vector,sources-internal_metrics,sources-host_metrics,transforms,sinks-blackhole,sinks-console,sinks-file,sinks-http,sinks-socket,sinks-vector'`, I am running into libssl-dev dependency error which I had not got when compiling the entire list. Trying to figure out what I might be missing.

**Reply by jszwedko** [reply]
Ah, yes, you can try adding in `vendor-openssl` as another feature. That will use the a vendored copy of openssl rather than linking with your system openssl. That is used by default.

You may just want to include the `vendor-all` feature which will use vendored versions of all dependencies.

**Reply by bmallya** [reply]
Thanks, Let me give this a try.

**Reply by bmallya** [reply]
Apologies for the late reply. I was able to compile vector using the vendor-all feature. The binary now stands at 62MB.

**Reply by jszwedko** [reply]
Thanks for the report!
